File: p157-read4/Solution.cc
// Forward declaration of the read4 API.

int read4(char *buf);

class Solution {
  public:
    /**
     * @param buf Destination buffer
     * @param n   Maximum number of characters to read
     * @return    The number of characters read
     */

    int read(char *buf, int n) {
      int i = n/4;
      int result = 0;
      int tmp, r;
      char b[4];
      while(i--) {
        tmp = read4(buf + result);
        result += tmp;
        if (tmp < 4) return result;
      }
      tmp = n - result;
      if (tmp > 0) {
        r = read4(b);
        tmp = r < tmp ? r : tmp;
        r = 0;
        while(r < tmp) {
          buf[result++] = b[r++];
        }
      }
      return result;
    }
};

File: p157-read4/Solution.java
/* The read4 API is defined in the parent class Reader4.
 * int read4(char[] buf);
 */

public class Solution extends Reader4 {

  /**
   * @param buf Destination buffer
   * @param n   Maximum number of characters to read
   * @return    The number of characters read
   */

  public int read(char[] buf, int n) {
    int result = 0;
    int mod = n % 4;
    int times = n / 4;
    int tmp, i;

    char[] subBuf = new char[4];
    if (n <= 0) return result;
    while(times > 0) {
      tmp = read4(subBuf);
      i = 0;
      while(i < tmp) {
        buf[result++] = subBuf[i++];
      }
      if (tmp < 4) return result;
      times--;
    }

    if (mod <= 0) return result;

    tmp = read4(subBuf);
    tmp = tmp > mod ? mod : tmp;
    i = 0;
    while(i < tmp) {
      buf[result++] = subBuf[i++];
    }
    return result;
  }
}
